Core/CoreTests/Features/Files/View/FileList/FileListViewControllerTests.swift

```diff
@@ -271,6 +275,54 @@ class FileListViewControllerTests: CoreTestCase {
         XCTAssertEqual(controller.tableView.numberOfRows(inSection: 0), 0)
     }
 
+    func test_addFileButtonVisible_whenNotRestricted() {
+        controller = FileListViewController.create(
+            env: environment,
+            context: .currentUser,
+            path: "Folder A",
+            studentAccessInteractor: studentAccessInteractor
+        )
+
+        controller.view.layoutIfNeeded()
+        controller.viewWillAppear(false)
+
+        _ = controller.addButton.target?.perform(controller.addButton.action)
+        let sheet = router.presented as? BottomSheetPickerViewController
+        let titles = sheet?.actions.map { $0.title } ?? []
+
+        XCTAssertTrue(titles.contains("Add File"), "Add File should be visible when not restricted")
+        XCTAssertTrue(titles.contains("Add Folder"), "Add Folder should always be visible")
+    }
+
+    func test_addFileButtonHidden_whenRestricted() {
+        studentAccessInteractor.setRestricted(true)
+
+        controller = FileListViewController.create(
+            env: environment,
+            context: .currentUser,
+            path: "Folder A",
+            studentAccessInteractor: studentAccessInteractor
+        )
+
+        controller.view.layoutIfNeeded()
+        controller.viewWillAppear(false)
+
+        let expectation = expectation(description: "Restriction applied")
+
+        // Give time for Combine publisher to deliver on main queue
+        DispatchQueue.main.async {
+            expectation.fulfill()
+        }
+
+        wait(for: [expectation], timeout: 1.0)
```

Comment:
A less flaky approach is to not use the main queue in tests but rather than a synchronous scheduler. Search for `AnySchedulerOf<DispatchQueue>` and `DispatchQueue.test` on how to use it.